// nodeApp/static/asset/component/business/taggroup/taggroup.js
/*
* 标签卡片
* @params:
*   store: 数据源
*   urlkey: url的key值，默认为key
*   displaykey: 显示的key值，默认为name
*   itemclick: 单机item事件
*   isdelete: 是否可删除，默认为true
*/
require('./taggroup.scss');
var tpl = require('./taggroup.jade');

var TagGroup = Vue.extend({
    props: ['store', 'urlkey', 'displaykey', 'isdelete', 'itemclick'],
    template: tpl(),
    data: function(){
        return {
            store: this.store || [],
            urlkey: this.urlkey || 'url',
            displaykey: this.displaykey || 'name',
            itemclick: this.itemclick || function(){},
            isdelete: this.isdelete || true
        }
    },
    methods: {
        /*
        * 取值
        */
        getValue: function(){
            return this.store;
        },
        /*
        * 添加tag
        */
        add: function(rec){
            rec = rec || {};
            this.store.push(rec);
        },
        /*
        * 删除tag
        */
        delete: function(rec, index){
            this.store.splice(index, 1);
            this.$dispatch('delete', rec);
        },
        /*
        * 清空tag
        */
        empty: function(){
            this.store = [];
        }
    }
});

module.exports = Vue.component('vtaggroup', TagGroup);
