// nodeApp/static/asset/component/base/form/form.js
/*
* 依赖工具类
*/
var parse = require('webShareAssets/utils/parse');

/*
* 表单组件: form
* @params:
*   action: type is String, default is ''
*   type: 'post' || 'get', default is 'post'
*/

require('./form.scss');
var tpl = require('./form.jade');

var Form = Vue.extend({
    template: tpl(),
    props: {
        action: {
            type: String,
            default: ''
        },
        type: {
            type: String,
            default: 'post'
        }
    },
    data: function () {
        return {

        }
    },
    methods: {
        ajaxForm: function (option = {}) {
            if (option.before && option.before() === false) {
                return false;
            }
            var _self = this,
                $thisForm = $(this.$el),
                data = {};
            $thisForm.find('input').each(function (index, el) {
                var $this = $(el),
                    type = $this.attr('type'),
                    name = $this.attr('name'),
                    val = $this.val();
                if (type === 'radio') {
                    if ($this.is(':checked')) {
                        data[name] = $this.val();
                    }
                } else if (type === 'checkbox') {
                    if ($this.is(':checked')) {
                        data[name] ? data[name].push(val) : data[name] = [val];
                    }
                } else {
                    data[name] = $this.val();
                }
            });
            $.ajax({
                url: this.action,
                type: this.type,
                data: data,
                success: function (rtn) {
                    if (!rtn.succeeded) {
                        return parse.showError(rtn);
                    }
                    option.success && option.success(rtn);
                }
            });
        }
    }
});

module.exports = Vue.component('vform', Form);
