// nodeApp/static/asset/component/business/guide/guide.js
/*
* 引导组件
* @params:
*   isNeed: 引导开关，上线后一周内关闭，默认为true
*   isShow: 是否显示引导，默认为false
*   current: current当前显示的页，默认为0（第一页）
*   store: 数据源
*   key: cookie键值，默认为guide
*/
require('./guide.scss');
var tpl = require('./guide.jade');

var Guide = Vue.extend({
    props: ['store', 'key'],
    template: tpl(),
    data: function () {
        return {
            isNeed: true,
            isShow: false,
            current: 0,
            key: this.key || 'guide'
        }
    },
    methods: {
        init: function () {
            if (!this.isNeed) {
                return;
            }
            var hasLoaded = this.getCookie();
            if (!hasLoaded) {
                this.isShow = true;
            }
            this.setCookie();
        },
        loadNext: function () {
            this.current++;
        },
        /*
        * 隐藏引导
        */
        hide: function () {
            this.isShow = false;
        },
        /*
        * 获取cookie
        */
        getCookie: function (key) {
            key = key || this.key;
            return $.cookie(key);
        },
        /*
        * 设置cookie
        */
        setCookie: function (key) {
            key = key || this.key;
            $.cookie(key, 'true', {
                expires: 999,
                path: '/'
            });
        }
    },
    ready: function () {
        this.init();
    }
});

module.exports = Vue.component('vguide', Guide);
