// nodeApp/static/asset/component/base/tabpanel/tabpanel.js
/*
* 面板组件: tabpanel
* @params:
*       active：当前激活tab
*       list：子元素
*/


require('./tabpanel.scss');
var tpl = require('./tabpanel.jade');

/*
* mixin
*/
var BaseMixin = require('../../mixin/base');

var TabPanel = Vue.extend({
    mixins: [BaseMixin],
    template: tpl(),
    props : ['active', 'list'],
    data: function(){
        return {
           index: this.active
        }
    },
    methods: {
        init: function(){
            this.$body = $(this.$el);
            this.$tabs = this.$body.find('.tab');
        },
        initEvents: function(){
        },
        /*
        * 获取当前激活tab
        */
        getActiveTab: function(){
            return this.index;
        },
        /*
        * 设置当前激活tab
        */
        setActiveTab: function(index){
            index = index || this.index;
            if(index > this.list.length -1){
                index = 0;
            }
            this.index = index;
        },
        /*
        * tab点击事件
        */
        tabClick: function(index){
            this.index = index;
            this.$dispatch('tabchange', index);
        },
        /*
        * 设置tab的title
        */
        setTitleByIndex: function(title, index){
            this.$tabs.eq(index || 0).find('.tab-title').text(title);
        }
    }
});

module.exports = Vue.component('vtabpanel', TabPanel);
