// nodeApp/static/asset/component/base/radiogroup/radiogroup.js
/*
* 面板组件: radiogroup
* @params:
*       layout: vbox/hbox，默认为hbox(横向排列)
*       hidelabel: 是否隐藏label，默认为false
*       fieldlabel: label文案
*       labelwidth: label宽度
*       defaultvalue: 初始化时选中的值，默认为第一项
*       items：radio单元，json数组：
*              [{
*                   value: '1',
*                   text: 'One' 
*              }]
*/


require('./radiogroup.scss');
var tpl = require('./radiogroup.jade');

/*
* mixin
*/
var FormMixin = require('../../mixin/form');

var RadioGroup = Vue.extend({
    mixins: [FormMixin],
    template: tpl(),
    props : ['layout', 'items', 'defaultvalue', 'hidelabel', 'fieldlabel', 'labelwidth'],
    data: function(){
        return {
            cls: this.layout === 'vbox' ? 'v-box':'',
            items: this.items || [],
            val: this.defaultvalue || this.items[0].value,
            isshowlabel: !this.hidelabel,
            labeltext: this.fieldlabel || '类型',
            labelstyle: {
                width: this.transformCssUnit(this.labelwidth || '200px')
            }
        }
    }
});

module.exports = Vue.component('vradiogroup', RadioGroup);
