// nodeApp/static/webpack.config.js
var path = require('path');
var config = require('./config');
var webpack = require("webpack");
var ExtractTextPlugin = require("extract-text-webpack-plugin");
// var AssetsPlugin = require('assets-webpack-plugin');
// var HtmlWebpackPlugin = require('html-webpack-plugin');

module.exports = {
    // 入口
    entry: {
        userIndex: config.assetDir + '/page/user/userIndex/userIndex',
    },
    // 输出
    output: {
        path: path.join(__dirname, config.tmpDir),
        filename: 'js/[name].js',
        publicPath: config.tmpDir
    },
    module: {
        // 加载器
        loaders: [{
            // 使用Babel转换ES6，排除node_modules目录下的js
            test: /\.js$/,
            loader: 'babel',
            exclude: /node_modules/
        }, {
            test: /\.css$/,
            loader: ExtractTextPlugin.extract('style-loader', 'css-loader')
        }, {
            test: /\.(scss|sass)$/,
            loader: ExtractTextPlugin.extract('style-loader', 'css-loader!sass')
        }, {
            test: /\.(png|jpg|gif)$/,
            loader: 'url-loader'
        }, {
            test: /\.jade$/,
            loader: "jade"
        }, {
            test: /\.(html|cshtml|tpl)$/,
            loader: 'html-loader'
        }]
    },
    resolve: {
        // require时省略的扩展名，如：require('module') 不需要module.js
        extensions: ['', '.js'],
        // 别名
        alias: {
            filter: path.join(__dirname, config.assetDir + '/filters'),
            components: path.join(__dirname, config.assetDir + '/components'),
            asset: path.join(__dirname, config.assetDir)
        },
        modulesDirectories: [path.join(__dirname, config.assetDir + '/etc')]
    },
    // 开启source-map，webpack有多种source-map，在官网文档可以查到
    // devtool: '#source-map',
    devtool: config.sourceMap,
    plugins: [
        new ExtractTextPlugin("css/[name].css", {
            allChunks: true
        }),
        //new webpack.optimize.UglifyJsPlugin({
        //    compress: {
        //      warnings: false
        //    }
        //})
    ],
    sassLoader: {
        includePaths: [path.join(__dirname, config.assetDir + '/etc')]
    }
};
