// nodeApp/static/asset/component/business/bread/bread.js
/*
* 面包屑组件
* @params:
*   breads: 面包屑元素，格式：
*           "我的首页 || /,邀请猎头 || /Headhunting/Invitations"
*           ==> 
*           "a(href="/")我的首页 a(href="/Headhunting/Invitations")邀请猎头 "
*/
require('./bread.scss');
var tpl = require('./bread.jade');

var Bread = Vue.extend({
    props: ['breads'],
    template: tpl(),
    data: function(){
        var bread = this.getBreads();
        return {
            vbreads: bread,
            count: bread.length - 1
        }
    },
    methods: {
        /*
        * 默认面包元素
        */
        getDefault: function(){
            return [{
                url: '/',
                text: '首页'
            }];
        },
        /*
        * 解析面包元素
        */
        getParse: function(arr){
            var item, bread = [];
            for(var i = 0, len = arr.length; i < len; i++){
                item = arr[i].split('||');
                bread.push({
                    text: item[0],
                    url: item[1]
                });
            }
            return bread;
        },
        getBreads: function(){
            var bread = [], arr = this.breads.split(',');
            if(arr.length === 0){
                bread = this.getDefault();
            }else{
                bread = this.getParse(arr);
            }
            return bread;
        }
    }
});

module.exports = Vue.component('vbread', Bread);
