// nodeApp/static/asset/component/mixin/list.js
/*
* mixin - list表单相关
*/

var BaseMixin = require('./base');
var Class = require('../core/class');

var List = {
    methods: {
        initEvents: function(){
            var _self = this;
            this.pager.$on('pager', function(data){
                _self.params.start = data * _self.vpagesize;
                _self.load();
            });
        },
        parseParams: function(){
            var url = '?';
            for(var i in this.params) {
                url += '&' + i + '=' + this.params[i];
            }
            return url;
        },
        addRows: function (records) {
            records = records || [];
            for(var i = 0, len = records.length; i < len; i++){
                var $row = this._createRow(records[i], i);
                this.body.append($row);
            }
        },
        _createRow: function(record, index){
            record = this.beforeRowRender(record);
            var row  = this.rowTpl({
                data: record,
                columns: this.columns
            });
            var $row = $(row);
            if(index % 2 !== 0){
                $row.addClass('active');
            }
            $row.data('record', record);
            this._bindEvents($row, record);
            return $row;
        },
        _empty: function(){
            this.body.html('');
        },
        _bindEvents: function(row, record){
            var _self = this;
            // row click
            row.on('click', function(){
                _self.$dispatch('rowclick', row, record);
            });
            row.find('.btn').on('click', function(){
                var $this = $(this);
                if($this.attr('disabled')){
                    return;
                }
                _self.$dispatch($this.data('action'), $this, record);
            });
        }
    }
};

Class.extend(BaseMixin, List);

module.exports = List;
